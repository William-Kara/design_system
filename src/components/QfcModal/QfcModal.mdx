import { ArgTypes, Meta } from '@storybook/blocks';
import { Story } from '@storybook/addon-docs';
import QfcModal from '@/components/QfcModal/QfcModal.vue';

<Meta title="Modules/Modal" />

# Definition

This is a QfcModal module that is used to display a modal with an image, title, and content after clicking a button. It's using QfcBox and QfcButton.

## Props

The following props are available for the QfcModal component:

<ArgTypes of={QfcModal} />

# Usage

To use QfcModal, style needs to be imported separately.

```sass

@import '@website/design-system-vuejs/dist/styles/bulma/components/modal.scss';
@import '@website/design-system-vuejs/dist/styles/components/QfcModal.scss';
```

```html
<template>
  <QfcModal v-model:isOpen="isOpen" title="Lorem" content="lorem ipsum" icon="/myImg.svg" />
</template>
```

Modal can also be used with template slots, allowing you to customize the title and content with html.You can also use the v-model:isOpen prop to control the modal's state.

```html
<template>
  <QfcModal v-model:isOpen="isOpen">
    <template #title>Title</template>
    <template #content>Content</template>
  <QfcModal>
</template>

<script>
 const isOpen = ref(false);
</script>
```

Clicking outside of the modal will allow you to close it. You can also close it by clicking on the close button.

# Exemple

<Story name="Primary" id="components-modal--primary" />
