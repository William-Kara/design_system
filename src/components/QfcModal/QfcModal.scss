.q-modal-card {
  background-color: $black;
  text-align: center;
  &__close {
    position: absolute;
  }
  &__content {
    max-width: 80%;
    margin: 4rem auto;
    display: flex;
    flex-direction: column;
    gap: $q-gap-l;
  }
  &__img {
    margin: auto;
  }
  &__title {
    font-weight: $semibold;
  }
  &__txt {
    font-size: $fs-s;
    font-weight: $regular;
  }
}
