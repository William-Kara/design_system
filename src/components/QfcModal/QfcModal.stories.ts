import Modal from './QfcModal.vue';
import type { Meta, StoryObj } from '@storybook/vue3';

const meta: Meta<typeof Modal> = {
  component: Modal,
  title: 'Modules/Modal',
  argTypes: {
    img: {
      type: { name: 'string', required: false },
      defaultValue: '/logo.svg',
    },
    title: {
      type: { name: 'string', required: false },
      defaultValue: 'Félicitations, vous êtes inscrit pour devenir testeur !',
    },
    content: {
      type: { name: 'string', required: false },
      defaultValue:
        'Nous vous remercions pour votre inscription. Vous serez informés par e-mail de la démarche à suivre pour tester nos formations et notre plateforme dans les mois à suivre.',
    },
    isOpen: {
      type: { name: 'boolean', required: true },
    },
  },
};
export default meta;

type Story = StoryObj<typeof meta>;

export const Primary: Story = {
  args: {
    isOpen: true,
  },
};
