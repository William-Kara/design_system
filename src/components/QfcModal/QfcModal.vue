<template>
  <div role="dialog" class="modal" :class="{ 'is-active': isOpen }">
    <div class="modal-background" @click="toggleModal()"></div>
    <div class="modal-content">
      <QfcBox class="q-modal-card">
        <div class="q-modal-card__content">
          <slot name="close-button">
            <button
              class="q-modal-card__close modal-close is-large"
              aria-label="close"
              type="button"
              @click="toggleModal()"
            />
          </slot>
          <img class="q-modal-card__img" :src="img" width="168" height="18" alt="logo" />
          <h3 class="q-modal-card__title">
            {{ title }}
            <slot name="title" />
          </h3>
          <p class="q-modal-card__txt">
            {{ content }}
            <slot name="content" />
          </p>
        </div>
      </QfcBox>
    </div>
  </div>
</template>

<script lang="ts" setup>
import QfcBox from '@/components/QfcBox/QfcBox.vue';

interface Props {
  img?: string;
  title: string;
  content: string;
  isOpen: boolean;
}
const props = withDefaults(defineProps<Props>(), {
  img: '/logo.svg',
  isOpen: false,
});
const emit = defineEmits(['update:isOpen']);
const toggleModal = () => emit('update:isOpen', !props.isOpen);
</script>

<style lang="scss" scoped src="./QfcModal.scss" />
