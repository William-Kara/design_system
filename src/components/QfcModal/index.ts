import QfcModal from './QfcModal.vue';

export { QfcModal };
export default QfcModal;
