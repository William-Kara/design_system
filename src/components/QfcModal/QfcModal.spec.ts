import { it, describe, beforeEach, expect } from 'vitest';

import { shallowMount } from '@vue/test-utils';
import QfcModal from './QfcModal.vue';

let wrapper: any = null;

const propsValues = {
  img: '/logo.svg',
  title: 'Title',
  content: 'Content',
  isOpen: true,
};

beforeEach(() => {
  wrapper = shallowMount(QfcModal, {
    propsData: propsValues,
  });
});

describe('QfcModal Component', () => {
  it('should mount properly', () => {
    expect(wrapper).toBeTruthy();
  });

  it('should have correct props', () => {
    expect(wrapper.props()).toEqual(propsValues);
  });

  it('should render correctly', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
});
