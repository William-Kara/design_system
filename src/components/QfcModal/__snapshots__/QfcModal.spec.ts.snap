// Vitest Snapshot v1

exports[`QfcModal Component > should render correctly 1`] = `
"<div data-v-9477621e=\\"\\" role=\\"dialog\\" class=\\"modal is-active\\">
  <div data-v-9477621e=\\"\\" class=\\"modal-background\\"></div>
  <div data-v-9477621e=\\"\\" class=\\"modal-content\\">
    <qfc-box-stub data-v-9477621e=\\"\\" size=\\"fullwidth\\" padding=\\"medium\\" class=\\"q-modal-card\\"></qfc-box-stub>
  </div>
</div>"
`;
